README.rst

```diff
@@ -0,0 +1,760 @@
+  gem, pypi, etc. Required.
+- **namespace**: some name prefix such as a Maven groupid, a Docker image owner,
+  a GitHub user or organization. Optional and type-specific.
+- **name**: the name of the package. Required.
+- **version**: the version of the package. Optional.
+- **qualifiers**: extra qualifying data for a package such as an OS,
+  architecture, a distro, etc. Optional and type-specific.
+- **subpath**: extra subpath within a package, relative to the package root.
+  Optional.
+
+
+Components are designed such that they for a hierarchy from the most significant
+on the left left to the least significant components to the right.
+
+
+A `purl` must NOT contain a URL Authority i.e. there is no support for
+`username`, `password`, `host` and `port` components. A `namespace` segment may
+sometimes look like a `host` but its interpretation is specific to a `type`.
+
+
+Some `purl` examples
+~~~~~~~~~~~~~~~~~~~~
+
+::
+
+    bitbucket:birkenfeld/pygments-main@244fd47e07d1014f0aed9c
+
+    deb:curl@7.50.3-1?arch=i386&distro=jessie
+
+    docker:cassandra@sha256:244fd47e07d1004f0aed9c
+    docker:customer/dockerimage@sha256:244fd47e07d1004f0aed9c?repository_url=gcr.io
+
+    gem:jruby-launcher@1.1.2?platform=java
+    gem:ruby-advisory-db-check@0.12.4
+
+    github:package-url/purl-spec@244fd47e07d1004f0aed9c
+
+    go:google.golang.org/genproto#googleapis/api/annotations
+
+    maven:org.apache.xmlgraphics/batik-anim@1.9.1?packaging=sources
+    maven:org.apache.xmlgraphics/batik-anim@1.9.1?repository_url=repo.spring.io/release
+
+    npm:%40angular/animation@12.3.1
+    npm:foobar@12.3.1
+
+    nuget:EnterpriseLibrary.Common@6.0.1304
+
+    pypi:django@1.11.1
+
+    rpm:curl@7.50.3-1.fc25?arch=i386&distro=fedora-25
+    rpm:curl@7.50.3-1.fc25?arch=src
+
+(NB: some checksums are truncated for brevity)
+
+
+A `purl` is a URL
+~~~~~~~~~~~~~~~~~
+
+- A `purl` is a valid URL and URI that conforms to the URL definitions or
+  specifications at:
+
+  - https://tools.ietf.org/html/rfc3986
+  - https://en.wikipedia.org/wiki/URL#Syntax
+  - https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#Syntax
+  - https://url.spec.whatwg.org/
+
+- The `purl` components are mapped to these URL components:
+
+  - `purl` `type`: this is a URL `scheme`
+  - `purl` `namespace`, `name` and `version` components: these are
+    collectively mapped to a URL `path`
+  - `purl` `qualifiers`: this maps to a URL `query`
+  - `purl` `subpath`: this is a URL `fragment`
+  - In a `purl` there is no support for a URL Authority (e.g. NO
+    `username`, `password`, `host` and `port` components).
+
+- Special URL schemes as defined in https://url.spec.whatwg.org/ such as
+  `file://`, `https://`, `http://` and `ftp://` are NOT valid `purl` types. They
+  may be used to reference URLs in separate attributes outside of a `purl` or in
+  a `purl` qualifier.
+
+- Version control system (VCS) URLs such `git://`, `svn://`, `hg://` or as
```

Comment:
I'm not sure why `github:` and `bitbucket:` would be valid but `git:` would not be (the same applies to the others)? github: is just a pointer to a source location on a named, proprietary, source control host, if that is valid I should also be able to have a point to a source location on *any* server that implements the Git (or alternate) protocol.

One argument could be that git://, svn://, hg:// do not refer to specific package type, but that's true of github: and bitbucket: as well, they're just a source location, at that source location could be anything.

Reply:
You have a good point here again!. Here are some reasons that come to mind(and this ties in not squatting existing URL/URI schemes and the thread on `Go`):

`git://`, `svn://` and `bzr://` and heck... even `cvs:`/`pserver:` to some extent are well known URL schemes (even though they might not be officially registered) so reusing then for a `purl` would lead to a certain level confusion IMHO

While you are right that `github:` and `bitbucket:` are pointers to named and proprietary source control hosts, here are some considerations:

- I do not want to take a stance on open vs. proprietary here even though I am clearly on the open side personally :heart_eyes: 

- they (`github:`, `bitbucket:` and `gitlab:` and likely a few more) are a bit more than than just VCS pointers: they expose an API, naming and other conventions, ticket trackers, release mechanisms, metadata and more that goes well beyond a bare `git` or `hg` repo. This grants them a package-status...

- ... and even more so, as they are even natively supported by several package managers as a place to "provision" code deps (such as in Go or PHP Composer)

- they are kinda prominent as a place to host package source code. 

- bare VCS URLs have their own purpose but there is no set naming conventions on what would be a package name. I may for instance infer that git://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git is the URL to Linus's Linux git tree but not much more than this. I'd be at loss to make the same guess on `git://git.kernel.org/pub/scm/linux/kernel/git/pjt/test.git`

So based on all this, as much as I love `git://`-like URLs -- so much so that I even helped spec these for [SPDX](https://github.com/spdx/spdx-spec/blob/cfa1b9d08903befdf03e669da6472707b7b60cb9/chapters/3-package-information.md#37-package-download-location-) -- I think they might be best left out as valid `purl` `types` and tracked either as a separate data item outside of a `purl` OR  use them as one of the `qualifiers` as [proposed](https://github.com/package-url/purl-spec/pull/1/files#diff-88b99bb28683bd5b7e3a204826ead112R683) unless you strongly object to this.

Could there be use cases that would be made impossible by using this approach?

Reply:
The scenario I was thinking originally is that in NPM's package.json you can depend on a package on any arbitrary git server, not just GitHub, and I'm not sure how I would represent that in a purl based on the current spec. The three pieces of information that would need to be encoded in the purl are that it is NPM, that it's accessible over the git protocol, and the location of it.

It's almost as if to support this scenario (and maybe it's a specialized case), you would need to do something like:
`npm+git:host.com/path/to/repo.git`

Reply:
The same applies to Python pip requirements and Gemfiles AFAIK.
I think this would come out this way:

`npm:mypackage@12.4.5?vcs_url=git://host.com/path/to/repo.git`

Reply:
or 
`npm:mypackage@12.4.5?vcs_url=git://host.com/path/to/repo.git@4345abcd34343`

Reply:
Actually we could define a sub mini-spec for a `vurl` aka. `version control URL` which could be similar to the SPDX location spec, but abstract the strict requirement to set the transport because it does not matter too much whether you use https, git or or ssh transports from a repo reference standpoint: this is still the same repo to me....
May be something like `vurl+git://host.com/path/to/repo.git@commitish#subpath/in/repo` ? this is still a strongly hierarchical URL (e.g. host  comes before version/commitish, and a subpath may exist only in a certain commitish or branch )
This could be somewhat redundant with SPDX and possibly https://maven.apache.org/scm/scm-url-format.html too.... though may be simpler and more pragmatic?

Just a thought :) I am not convinced this is needed

Reply:
You'd definitely want to have some way of representing a generic Git or other VCS repository. There are still a lot of people who host their own VCS repositories (me included!).

I don't think it's viable to try and add every single VCS host on earth; there's no central catalog of these, and adding them as-they-are-encountered would create an ongoing workload and a lot of delays (as it will take time for individual hosts to be added), and disincentivize use of this specification for people who can't afford to wait for those host additions.

Simply having a `?vcs` qualifier doesn't quite cut it, either - sometimes the VCS repository is the *only* place in which a given piece of software is published, and there are no packages for it published elsewhere.

Reply:
@joepie91 thank you  for chiming in, you wrote:

> Simply having a ?vcs qualifier doesn't quite cut it, either - sometimes the VCS repository is the only place in which a given piece of software is published, and there are no packages for it published elsewhere.

But even if these cases, these are still packages, right? e.g. this would still be an NPM with a proper package.sjon and a name and version, correct?

Reply:
@pombredanne No, it would not necessarily be. The only moment where a package name in package.json is checked for uniqueness is when you try to actually *publish* it - if I wanted to, I could create a package named `express` and simply never publish it to NPM, and that would work fine when used locally.

In that scenario, the primary (and only unique) identifier for that 'package' would be its Git repository - the package.json *does* have a `name` in it, but that name actually belongs to a *different* package from a registry perspective (and basing an identifier on that would therefore result in the identifier describing the wrong package).

This applies even moreso for eg. languages that don't *have* a package registry or even a package manager at all, where a VCS repository may well be the only place to obtain the package from.

Reply:
You are making good points but the intent of this spec is to cater first and foremost to the common case, e.g. a well known published package: this should be clean, simple and concise. The simple case should be simple. It should be possible to cater to the complicated, goof and mistakes or unpublished things. That may not be simple then, just possible.

The goal is that when @andrew libraries.io and @jayfk pyup and @ashcrow and @jasinner victims and @Grafeas and my https://github.com/nexB/scancode-toolkit and more can all talk about the same package when we talk about about the same simple `purl` 

So in the scenario you list above, you have a goofy messed up case. A `purl` cannot make this any less goofy. In this case it might be something such as `npm:express@12.2?vcs_url=yourproprietaryrepo.git@commit` which may be incorrect name- and version-wise but that is still what is this package is known as when you install it with `npm` or `yarn`. 
The same would apply to an un-named private composer package. Weird and goofy, legit for composer but not something we should model a spec after IMHO and certainly not the common case. And in the vast majority of the cases, these weird and goofy things may not even need a purl at all.

> This applies even moreso for eg. languages that don't have a package registry or even a package manager at all, where a VCS repository may well be the only place to obtain the package from.

That would be the case of Golang with `go get` but even there is a name alright and an explicit origin in the namespace. 

Is there any public package repository that provides no name for its packages and only a version control system instead?    None that I can think of.

And even if this is so we can always find a warty way to accommodate warts and not make the wart the norm for the common case.

Reply:
@joepie91 reading the package.json [spec](https://docs.npmjs.com/files/package.json#name):

> The most important things in your package.json are the name and version fields. Those are actually required, and your package won't install without them. The name and version together form an identifier that is assumed to be completely unique. Changes to the package should come along with changes to the version.

So yes, one could create goofy things, but this would break the central assumption of the `npm` "protocol" that `The most important things in your package.json`.

Again, fine by me, but I do neither have to care for this, nor cater to this, would you agree?

Reply:
I fear you're moving the goalpost here. Your original premise is that a "`purl` is a URL string used to identify and locate a software package in a mostly universal and uniform way". This makes sense, for that which are clearly software packages: npm, gem, rpm, and so on. But by including `github`, you make the implicit claim that a repository on Github can be seen as a software package.

And if you make this claim, then it's my belief that you need to extend that to cover any Git vcs. From what can be seen as defining a "software package", I see no discerning difference between one hosted on Github and one hosted elsewhere.

Else, this all looks very good to me :-)

Reply:
@jonasob Thank you ++ for your feedback. Much honored!

You wrote:

> And if you make this claim, then it's my belief that you need to extend that to cover any Git vcs. From what can be seen as defining a "software package", I see no discerning difference between one hosted on Github and one hosted elsewhere.

I am not making an implicit claim about a repo on Github. Instead this is more about acknowledging that the practice is mostly to treat these as such in @andrew and @benjam [Libraries.io](https://github.com/librariesio/libraries.io/tree/5213d3b03354aed98bbb36c2db92475d093f6e76/app/models/concerns), in @reiz [versioneye](https://github.com/versioneye/crawl_r/tree/master/lib/versioneye/crawlers), in @jpopelka and @fridex [fabric8](https://github.com/fabric8-analytics/fabric8-analytics-worker/blob/aea81992d419e2c984c3962cbd775590d52a5a08/f8a_worker/utils.py#L34) and a few more, including my own code.  

This is just recognizing that a Kalithea, GitHub, Savanah, Sourceforge, Gitlab, Bitbucket, etc "repo" is more than just a VCS repo but comes with a package-like set of data e.g. a namespace and naming conventions, licensing & provenance data, releasing, APIs, tickets, docs, etc. similar to the data accumulated in package manifests. 

In contrast a bare VCS repo does not have much of these. It is however an important (I'd say essential) locator in many cases. So here is how I would handle a bare Git repo:

`generic:<somename>?vcs_url=git://myhost/mypath/myrepo.git@commitish` 

<somename> could be at worst the name of the repo.
(adding if needed a subpath to that vc_url) using essentially the same URL convention that I contributed to SPDX. We could even relax this URL syntax a bit and define it clearly in this spec (for instance the fact we use the git, ssh, or https protocol is not of much interest compared to the fact this is a git or hg repo.). 

See also this related [discussion](https://github.com/package-url/purl-spec/pull/1#discussion_r151830997)
 
> Else, this all looks very good to me :-)

Thanks!

Reply:
@jonasob So for instance, I would consider [gitea](https://git.fsfe.org/jonas)  that you use at the FSFE as a `type` as it has all the same attributes as a Bitbucket and similar.

Reply:
What does `npm:mypackage@12.4.5?vcs_url=git://host.com/path/to/repo.git` refer to exactly? Because npm specifiers can be a version OR a git URL, and git URLs can have versions, eg, the following are valid and each refer to a different thing:
```
mypackage@12.4.5
mypackage@git://host.com/path/to/repo.git
mypackage@git://host.com/path/to/repo.git#semver:12.4.5
```
Further, if you have `mypackage@12.4.5` then you can get the vcs URL from the package metadata.

npm supports `bitbucket:`, `github:`, `gitlab:` and `gist:` prefixes. (Implementation is [here](https://github.com/npm/hosted-git-info#readme))